Add Filament Manager Integration

techfreek opened the pull request:
I use the Filament Manager plugin to track filament usage, and I've found when using OctoScreen taht I often forget to swap the filament because I have to go back to octoprint to change the selected filament.

Anyways, I figured adding an integration for a plugin wouldn't be too difficult, so I through together an integration

Adds:
* Button to the existing Filament panel to navigate to the Filament Manager panel (if plugin is installed)
* Adds Panel to view the Filament Manager status of all tool heads
* Adds Panel to view/set the filament per toolhead
  * If only 1 toolhead is present, the first panel is skipped
* The relevant octoprintApis/dataModels for the 3 API calls we need to make
* Adds a utility to get plugin information (basically wrapping PluginManagerInfoRequest to get info on a single plugin)

Let me know if you have any thoughts or feedback

Examples of the 2 new panels and updated filament panel
![PXL_20210818_154048244](https://user-images.githubusercontent.com/3945675/129928720-51d8c61a-a477-4fa5-b5fb-b2511e75bc2a.jpg)
![PXL_20210817_062602698](https://user-images.githubusercontent.com/3945675/129928741-156d7704-ad5d-408d-ba00-5bd852c069ad.jpg)
![PXL_20210818_032706010](https://user-images.githubusercontent.com/3945675/129928756-0f1cc5f6-1cd6-41f4-a45e-a984d55513a3.jpg)

TODO:

- [ ] Add colored bar under Filament Manager Button
- [ ] Remove spool color underline
- [ ] Fix 1 spool display bug
- [ ] Make full row clickable

techfreek committed:
Add Filament Manager Integration

I use the Filament Manager plugin to track filament usage, and I've found when using OctoScreen taht I often forget to swap the filament because I have to go back to octoprint to change the selected filament.

Anyways, I figured adding an integration for a plugin wouldn't be too difficult, so I through together an integration

Adds:
* Button to the existing Filament panel to navigate to the Filament Manager panel (if plugin is installed)
* Adds Panel to view the Filament Manager status of all tool heads
* Adds Panel to view/set the filament per toolhead
  * If only 1 toolhead is present, the first panel is skipped
* The relevant octoprintApis/dataModels for the 3 API calls we need to make
* Adds a utility to get plugin information (basically wrapping PluginManagerInfoRequest to get info on a single plugin)

JeffB42 commented:
Thanks for contributing and submitting the PR.  I'll try to take a look at it next weekend.

techfreek commented:
Thanks for the developing this project! I really enjoy it.

techfreek commented on ui/FilamentPanel.go:
```diff
@@ -109,17 +111,50 @@ func (this *filamentPanel) initialize() {
+		column++
+	}
+
+	if utils.FilamentManagerPluginIsInstalled(this.UI.Client) {
```
If octoprint isn't running when the filament panel is initialized, we won't add the filament manager button unless we restart Octoscreen.

techfreek committed:
Include spool material in the details pane

The selected spool didn't show any information about the type of material that was loaded. So you couldn't tell if it was PLA/PETG/etc. Include in the details tab

techfreek committed:
Allow the FilamentPanel to update

On each load of the FilamentPanel, redraw the panel. This will ensure that if the page is loaded before OctoPrint has loaded, we can still correctly determine if the FilamentManager plugin is installed

JeffB42 commented:
@techfreek I apologize and am very sorry for taking so long to get to this.  I've been extremely busy the last several months, and finally got around to taking a look at your PR.

First, I want to say thanks for contributing.

Next, I haven't gone through the code in the PR yet. Instead, I smoke tested your branch, and I have a few notes:


* Since clicking on the Filament Manager navigates to a new panel, there should be a colored bar underneath it:
![image](https://user-images.githubusercontent.com/10328858/136728905-b8640bb3-46c0-4750-9065-51dbdef294ce.png)
Take a look at the Temperature button as an example.  For reference, there's https://github.com/Z-Bolt/OctoScreen/wiki/Version-2.6#user-interface-and-user-experience-improvements (go this link, and scroll to _"...when you see a button with a solid bar underneath it"_



* When I have only one spool configured, what's displayed looks wrong:
![image](https://user-images.githubusercontent.com/10328858/136728785-7cc1447e-19d4-48b3-b47d-e2c67776db57.png)

* I next added two more spools (total of three) and I see this:
![image](https://user-images.githubusercontent.com/10328858/136728824-35440195-6012-4250-b073-f1e6372936e5.png)

This matches the screenshot you provided, but I have some comments on it:
** "Tool0" - Please use "Extruder 1" instead.  There's a comment about moving away form "Tool0" and being 1-based, and using "Hotend 1" and "Extruder 1" in the v2.6 release notes:
https://github.com/Z-Bolt/OctoScreen/wiki/Version-2.6#user-interface-and-user-experience-improvements

** I haven't thought this fully through, but I think you'll need to support
a) machines with a single extruder and single nozzle
b) machines with a multiple extruders and single nozzle
c) machines with a multiple extruders and multiple nozzles 
If you have any questions about this, let me know and I can provide more details on this.

** There are several images with colored bars underneath them, but they appear to be decorative only and break the UI style of the app.  See the 2.6 release notes - an image with a yellow bar means user input will be required, and images with any other color bar means the user will be taken to a new panel.  I understand the reason they were added, but since clicking on a spool with a colored bar doesn't navigate to a different panel, this breaks the continuity of the UI.

** I'm not able to click on the entire row.  I can click on the spool image to select, but not the entire row.  For an example of how to do this, look at `addItem()` in https://github.com/Z-Bolt/OctoScreen/blob/master/ui/FilesPanel.go

techfreek commented:
Hey no worries about taking a while to take a look. I get it. Just responding with some discussion about some of the points before I start implementing changes.
> Since clicking on the Filament Manager navigates to a new panel, there should be a colored bar underneath it:

Yep, totally makes sense. I felt something was off there. That's an easy fix.

> When I have only one spool configured, what's displayed looks wrong:

I hadn't seen this one, I'll look into it. I definitely agree that it looks wrong.

> ** "Tool0" - Please use "Extruder 1" instead. There's a comment about moving away form "Tool0" and being 1-based, and using "Hotend 1" and "Extruder 1" in the v2.6 release notes:

This is one aspect of integrations that becomes difficult. I went for "Tool 0" as that is what I saw in the Filament Manager UI:
![image](https://user-images.githubusercontent.com/3945675/136895881-8cf68a0d-f9aa-4a15-a1d7-96b0ac2c431a.png)

I'm totally onboard with standardization, but that's just one place where it seemed like it might be hard to have consistency on both sides. I'm assuming you'll want to keep with the Tool 1 for consistency, but let me know if you have other thoughts. 

> ** I haven't thought this fully through, but I think you'll need to support
> a) machines with a single extruder and single nozzle
> b) machines with a multiple extruders and single nozzle
> c) machines with a multiple extruders and multiple nozzles

Totally agree. The way that I implemented the Filament Manager button is by having it check how many tools the Filament Manager plugin is reporting. If it reports a single tool, the button will shortcut directly to that specific tool. Otherwise, it will go to an intermediate window showing the tools as reported by Filament Manager, then you can select which tool and go to the same menu seen above. 

Example:

![image](https://user-images.githubusercontent.com/3945675/136897836-3c4feb20-2778-40d3-9a7a-3a9f16bbfe2b.png)


Admittedly, while fetching that screenshot I found a bug in that UI where I don't refresh the content. I'll definitely add that to my TODO list (I'll update the description with a todo list of open items. Some things I'll wait until I hear back before I add them).

> ** There are several images with colored bars underneath them, but they appear to be decorative only and break the UI style of the app. See the 2.6 release notes - an image with a yellow bar means user input will be required, and images with any other color bar means the user will be taken to a new panel. I understand the reason they were added, but since clicking on a spool with a colored bar doesn't navigate to a different panel, this breaks the continuity of the UI.

Understandable. If I find a different way to add a color to the spool (that isn't underlining it), would that be okay? Or should I just nix it entirely?

> ** I'm not able to click on the entire row. I can click on the spool image to select, but not the entire row. For an example of how to do this, look at addItem() in https://github.com/Z-Bolt/OctoScreen/blob/master/ui/FilesPanel.go

Yeah.... I know. I don't know how many hours (or was it days) I spent looking at how the Files Panel was implemented, and why the UI I was building kept getting horribly formatted whenever I tried to link the entire row. I'll definitely give it another try. Maybe not looking at it for a while will be what I needed to figure it out.

razserv2010 commented:
that cool, how do i add this? 
i did fresh install to octoscreen but this part is missing..

Codex- commented:
@razserv2010 that's because this is a Pull Request, it has not been merged into the projects main code yet.

JeffB42 commented:
> @razserv2010 that's because this is a Pull Request, it has not been merged into the projects main code yet.

@techfreek , any update on this PR?  Also, when ready, please target the 2.8.0-dev branch.

techfreek commented:
@JeffB42 I was waiting for answers to a couple questions I had above to make sure I didn't waste effort. Repeating/rewording them:
* Tool 0 vs Extruder 1. This was kind of a no good option since FilamentManager is 0 indexed, and OctoScreen is 1 indexed. I figure we'll end up going with 1 indexed, but I'm not sure how to communicate the discrepancy in indexing. Let me know if you have thoughts
* I know the idea of having a color underline for the different spools (to highlight different colors) clashes with the design language of OctoScreen, but I wasn't sure if you didn't like the idea of color coding or just the existing implementation (if you have a suggestion of how to do this in a way that works with the design language, let me know)
* I'm not sure what you'd like to see in terms of supporting different extruder/nozzle counts. This is attempting to just be a presentation layer for the FilamentManager plugin, so if it handles it, the OctoScreen plugin should handle it too

JeffB42 commented:
Sorry, I didn't realize you were waiting on me.

* I was just looking at the Filament Manager's UI, and I see what you mean (with it saying, "tool 0").  I don't think there's a good solution to this (damned if you do/damned if you don't).  I'd like to go with 1-based.

* yea, the color bar breaks the UI convention.  I was thinking, instead of a colored bar underneath the icon, there could be colored version of the icons, but I don't like that route either - that means the same icon would need to be duplicated for each and every color supported.  I've come up with two options:

* Option A) Instead of a color bar underneath the icon, set the color of the text to that color, like so:
![image](https://user-images.githubusercontent.com/10328858/160311443-f476f0f0-c30e-4ff7-895e-635a03376cbc.png)

* Option B) Instead of a color bar underneath the icon, add a "Color:" label, and a square swatch of the color, like so:
![image](https://user-images.githubusercontent.com/10328858/160311739-97817896-2948-41f6-8941-c98d00550655.png)

With option A, one issue would be if the color is black, or gray.

* As for multiple extruder support, I'm not sure what Filament Manager does.  Two configurations come to mind:
- supporting single nozzle, multiple extruder systems, like Prusa's Multi Material, or the Mosaic Palette 
You can simulate this by going to OctoPrint > Settings > Printer Profiles > Edit > Hot end & Extruder, and set the number of extruders to 2 (or more) and shared nozzle to "on"

- supporting multiple nozzle, multiple extruder systems, like the CoreXY machines with nozzle changers.
You can simulate this by going to OctoPrint > Settings > Printer Profiles > Edit > Hot end & Extruder, and set the number of extruders to 2 (or more) and shared nozzle to "off"

With a single nozzle/multiple extruders setup, my guess/hope is that the Filament Manager handles this.

With a multiple nozzles/multiple extruders setup, my guess is that you will need to call `FilamentManagerSetSelectionUri` for each tool that's present.  I've only looked through your PR, and I could have missed it, but it didn't look like it was accounting for multiple extruders.

* One other note to mention, could you change the target branch to 2.8.0-dev?  This is the active branch and will be used for the next release.  If you run into any issues building 2.8.0-dev, please let me know.  I've made changes to the branch and not everything is checked in.  I've made a lot of changes to main.go and ui.go but haven't checked them in yet, but I think what's check in at the moment builds & runs.  If you run into any build issues, please let me know and I'll get it working.

If these issues get address, I'll merge the PR in and look into issues like the screen resizing.

Thanks!

Jeff

Codex- commented on styles/z-bolt/style.css:
```diff
@@ -125,6 +125,43 @@ button.color-none {
 	border-bottom: 8px solid #E6D300;
 }
 
+.color-black {
+	border-bottom: 8px solid black;
+}
```
Is this used? There is a comment below stating `// skip setting for black since black on black won't be seen` and this class doesn't seem to be returned anywhere.

Codex- commented on ui/FilamentManagerToolPanel.go:
```diff
@@ -0,0 +1,328 @@
+}
+
+
+func getSpoolColor(spool *dataModels.FilamentManagerSpool) string {
```
instead of returning the name of a CSS class that has been defined already and having to write/generate a CSS class for each colour, couldn't this return a string literal of the CSS defining the colours class 🤔

Codex- commented on octoprintApis/dataModels/FilamentManagerSelections.go:
```diff
@@ -0,0 +1,16 @@
+
+type FilamentManagerSelections struct {
+	Selections []*FilamentManagerSelection `json: "selections"`
+}
```
nit: most of the repository has an EOF NL, but some of these new files are missing some

Codex- reviewed (commented):
My opinion doesn't really count for anything here, I'm just a fan :)

You could also try tinting the spool icon itself with the selected colour, I like the thinking of following the style set already throughout the UI with the bar, I would space the bar away from the spool, anchored to the bottom of the spool icon box and 100% of the boxes width like it's used elsewhere.

Having a label which says `Colour: blue/black` is great for accessibility, colour blindness isn't uncommon

JeffB42 commented:
Just a heads up on this branch.  The current development branch (2.8.0-dev) has deviated a bit from the master branch.  I merged the changes into my local but there were lots of conflicts.

After resolving the conflicts, I found lots os UX issues (like some of the ones I pointed out before).  My next PR will be in a week or so, and will incorporate some of the changes here, but with a little makeover.  I'm going to keep this PR open, and will close it once my next changes get merged in.

JeffB42 commented:
Hi @techfreek, thanks for your contribution.  I've finished the impl of FilamentManager support in the 2.8.0-dev branch, and I'm going to close this PR.  (normally I would take the PR, but the 2.8 branch had a lot of changes in it, and I felt it would be easier to add the changes myself, plus I made quite a bit of changes to get the feature to support multiple extruders, and not expand the screen).  If curious, here are some screen shots...


Here is the Filament panel, with the new Filament Manager icon
(note: this is also when the printer has multiple extruders)
<img width="398" alt="FilamentPanel" src="https://github.com/Z-Bolt/OctoScreen/assets/10328858/f44a98e6-0132-4775-80f4-6a65e54df5cc">

Here is the Filament Manager panel (when there is just a single extruder)
<img width="398" alt="FilamentManagerPanel-single extruder" src="https://github.com/Z-Bolt/OctoScreen/assets/10328858/b1ad8ca7-9873-4b46-a97b-bb81f8390a2a">

Here is the Filament Manager panel (when there are multiple extruders)
<img width="400" alt="FilamentManagerPanel-multiple extruders" src="https://github.com/Z-Bolt/OctoScreen/assets/10328858/f94f36ea-2741-404d-96b9-0d23b45aeba1">
